### PR TITLE
Eos 25484: UT for be allocator changes

### PR DESCRIPTION
**Pull request by ydb242**
# Problem Statement
- Write/Modify UT's to test and debug BE allocator changes

# Coding
   Checklist for Author
-  [x] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [x] Unit and System Tests are added
- [x] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [x] Testing was performed with RPM

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [x] Interface change (if any) are documented
- [x] Side effects on other features (deployment/upgrade)
- [x] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [x] JIRA number/GitHub Issue added to PR
- [x] PR is self reviewed
- [x] Jira and state/status is updated and JIRA is updated with PR link
- [x] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [x] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

File: be/ut/alloc.c [hunk 1]
```diff
@@ -398,6 +398,202 @@ M0_INTERNAL void m0_be_ut_alloc_spare(void)
...
+	M0_SET_ARR0(ut_ptr);
+
+	/**
+	 *  Alloc some chunks with chunk_align parameter set as true and
+	 *  remaining with chunk_align set as false.
+	 */
+	for (i = 0; i < ut_nr; i++) {
+		j = i;
+		ut_size  = m0_rnd64(&j) % BE_UT_ALLOC_SIZE + 1;
+		ut_tval  = i%2 == 0 ? true : false;
```
Comment (RadhaGulhane13) [inline be/ut/alloc.c:513]: space before and after %

File: be/ut/alloc.c [hunk 2]
```diff
@@ -398,6 +398,202 @@ M0_INTERNAL void m0_be_ut_alloc_spare(void)
...
+		ut_shift = i%2 == 0 ? BE_UT_ALLOC_SHIFT - 1 : BE_UT_ALLOC_SHIFT;
```
Comment (RadhaGulhane13) [inline be/ut/alloc.c:514]: same comment as above

File: be/ut/alloc.c [hunk 3]
```diff
@@ -398,6 +398,202 @@ M0_INTERNAL void m0_be_ut_alloc_spare(void)
...
+
+		M0_BE_UT_TRANSACT(ut_be, tx, cred,
+			  (m0_be_allocator_credit(a, M0_BAO_ALLOC_ALIGNED,
+						 ut_size, ut_shift, &cred),
```
Comment (RadhaGulhane13) [inline be/ut/alloc.c:518]: one extra space needed before ut_size

File: be/ut/alloc.c [hunk 4]
```diff
@@ -398,6 +398,202 @@ M0_INTERNAL void m0_be_ut_alloc_spare(void)
...
+	 *  Delete every third chunk to make sure that both type of chunks are
+	 *  deleted.
+	 */
+	for (i = 0; i < ut_nr; i+=3) {
+		j = i;
+		ut_size  = m0_rnd64(&j) % BE_UT_ALLOC_SIZE + 1;
+		ut_shift = i%2 == 0 ? BE_UT_ALLOC_SHIFT - 1 : BE_UT_ALLOC_SHIFT;
+		M0_BE_UT_TRANSACT(ut_be, tx, cred,
+			  (m0_be_allocator_credit(a, M0_BAO_FREE_ALIGNED,
+						 ut_size, ut_shift, &cred),
```
Comment (RadhaGulhane13) [inline be/ut/alloc.c:550]: one extra space needed before ut_size

File: be/ut/alloc.c [hunk 5]
```diff
@@ -398,6 +398,202 @@ M0_INTERNAL void m0_be_ut_alloc_spare(void)
...
+		else
+			M0_UT_ASSERT(m0_addr_is_aligned(ut_ptr[i],
+							BE_UT_ALLOC_SHIFT));
+	}
+
+	/**
+	 *  Delete every third chunk to make sure that both type of chunks are
+	 *  deleted.
+	 */
+	for (i = 0; i < ut_nr; i+=3) {
```
Comment (RadhaGulhane13) [inline be/ut/alloc.c:544]: space before, after +=

Comment (ydb242) [reply be/ut/alloc.c]: Done

File: be/ut/alloc.c (hunk 4 above)
Comment (ydb242) [reply be/ut/alloc.c]: Done

File: be/ut/alloc.c (hunk 1 above)
Comment (ydb242) [reply be/ut/alloc.c]: Done

File: be/ut/alloc.c (hunk 2 above)
Comment (ydb242) [reply be/ut/alloc.c]: Done

File: be/ut/alloc.c (hunk 3 above)
Comment (ydb242) [reply be/ut/alloc.c]: Done

File: be/ut/alloc.c [hunk 6]
```diff
@@ -398,6 +398,204 @@ M0_INTERNAL void m0_be_ut_alloc_spare(void)
...
+			   m0_be_alloc_stats_credit(a, &cred)),
+			  (M0_BE_OP_SYNC(op,
+					 m0_be_free_aligned(a, tx, &op,
+							    ut_ptr[i])),
+			   m0_be_alloc_stats_capture(a, tx)));
+		ut_ptr[i] = NULL;
+	}
+
+	/* Verify the alignment of the remaining chunks */
+	for (i = 0; i < ut_nr; i++) {
```
Comment (shashank-parulekar) [inline be/ut/alloc.c:471]: If you have deleted some allocations then you need to cut the iteration of this loop.

File: be/ut/alloc.c [hunk 7]
```diff
@@ -398,6 +398,204 @@ M0_INTERNAL void m0_be_ut_alloc_spare(void)
...
+						     M0_BITS(M0_BAP_NORMAL),
+						     ut_tval)),
+			   m0_be_alloc_stats_capture(a, tx)));
+		M0_UT_ASSERT(ut_ptr[i] != NULL);
+	}
+
+	/* Verify the alignment of the chunks */
+	for (i = 0; i < ut_nr; i++) {
+		if (i%2 == 0)
+			M0_UT_ASSERT(m0_addr_is_aligned(ut_ptr[i],
```
Comment (shashank-parulekar) [inline be/ut/alloc.c:534]: As mentioned above the alignment of the chunk header should be checked and not the allocated memory's start address.

File: be/ut/alloc.c [hunk 8]
```diff
@@ -398,6 +398,204 @@ M0_INTERNAL void m0_be_ut_alloc_spare(void)
...
+	 * 2. Verify the alignment of all the allocated chunks.
+	 * 3. Delete some of the allocated chunks.
+	 * 4. Verify the alignment of the remaining chunks.
+	 * 5. Delete remaining of the allocated chunks.
+	 */
+
+	M0_SET_ARR0(ut_ptr);
+
+	/**
+	 *  Alloc some chunks with chunk_align parameter set as true and
```
Comment (shashank-parulekar) [inline be/ut/alloc.c:507]: s/some chunks/half memory/

File: be/ut/alloc.c [hunk 9]
```diff
@@ -398,6 +398,204 @@ M0_INTERNAL void m0_be_ut_alloc_spare(void)
...
+							&cred),
+				 m0_be_alloc_stats_credit(a, &cred)),
+				(M0_BE_OP_SYNC(op,
+					       m0_be_free_aligned(a, tx, &op,
+								  ut_ptr[i])),
+				 m0_be_alloc_stats_capture(a, tx)));
+			ut_ptr[i] = NULL;
+		}
+	}
+
```
Comment (shashank-parulekar) [inline be/ut/alloc.c:592]: There is a function m0_be_allocator__invariant() in alloc.c. You can use it to make sure that everything is fine with the allocated memory.

**Review by shashank-parulekar — CHANGES_REQUESTED** [review]

File: be/ut/alloc.c [hunk 10]
```diff
@@ -398,6 +398,204 @@ M0_INTERNAL void m0_be_ut_alloc_spare(void)
...
+						     ut_size, ut_shift,
+						     M0_BITS(M0_BAP_NORMAL),
+						     true)),
+			   m0_be_alloc_stats_capture(a, tx)));
+		M0_UT_ASSERT(ut_ptr[i] != NULL);
+	}
+
+	/* Verify the alignment of the chunks */
+	for (i = 0; i < ut_nr; i++)
+		M0_UT_ASSERT(m0_addr_is_aligned(ut_ptr[i], ut_shift));
```
[the remark this answers is not in the document]
Comment (ydb242) [reply be/ut/alloc.c]: Done

File: be/ut/alloc.c (hunk 7 above)
Comment (ydb242) [reply be/ut/alloc.c]: Done

File: be/ut/alloc.c (hunk 9 above)
Comment (shashank-parulekar) [reply be/ut/alloc.c:600]: It is called by m0_be_alloc_aligned only when expensive checks are turned ON. Confirm if this is true when ut's are run and if-not then explicitly make a call to the invariant in your ut.

Comment (ydb242) [reply be/ut/alloc.c:606]: Added calls for m0_be_allocator__invariant at the start of the UT, after allocation of memory, after freeing up of memory and at the end of the UT.

**Review by RadhaGulhane13 — APPROVED** [review]

**Review by shashank-parulekar — APPROVED** [review]

**Review by NikhilBirgade — APPROVED** [review]

**Pull request merged**